### ISSUE TITLE
Failed notification step for passing builds results in subsequent failure notifications

### ISSUE THREAD

**Issue by log0ymxm** [drone/drone#1474]
This is happening with the drone 0.4 docker image. The hipchat notifier sends updates to our hipchat room, but the build status is always posted as failed, regardless of the actual build status.

**Comment by log0ymxm** [comment]
![screen shot 2016-02-09 at 10 51 57 am](https://cloud.githubusercontent.com/assets/35297/12925375/3ba1b604-cf1b-11e5-8595-4166cb5a14bd.png)
![screen shot 2016-02-09 at 10 52 05 am](https://cloud.githubusercontent.com/assets/35297/12925376/3d9b515e-cf1b-11e5-958a-f34228a71aca.png)

**Comment by bradrydzewski** [comment]
cc @jackspirou

**Comment by bradrydzewski** [comment]
@log0ymxm also can you please post a full yaml with any sensitive data redacted?

**Comment by log0ymxm** [comment]
```yaml
cache:
  mount:
    - venv
    - .git
compose:
  database:
    image: mysql
    pull: true
    environment:
      - <env_vars...>
    volumes:
      - /var/run/mysqld/mysqld.sock:/var/run/mysqld/mysqld.sock
build:
  image: log0ymxm/django-docker
  commands:
    - <build-steps...>
notify:
  email:
    recipients:
      - <recipients>
    on_success: false
    on_failure: blame
  hipchat:
    auth_token: <token>
    room_id_or_name: Build
    notify: true
    when:
      success: true
      failure: true
      change: true
    template: >
      <strong>{{ uppercasefirst build.status }}</strong> <a href=\"{{ system.link_url }}/{{ repo.owner }}/{{ repo.name }}/{{ build.number }}\">{{ repo.owner }}/{{ repo.name }}#{{ truncate build.commit 8 }}</a> ({{ build.branch }}) by {{ build.author }} in {{ duration build.started_at build.finished_at }} </br> - {{ build.message }} {{#success build.status}} It works {{/success}}
```

I've got the template in there just so I could debug. The `build.status` is always failed.

**Comment by jackspirou** [comment]
@log0ymxm taking a look at our system and the hipchat plugin to see if I can reproduce.

**Comment by jackspirou** [comment]
@bradrydzewski The hipchat is working fine for us internally, on the latest hipchat plugin code.  The last time I updated our drone to the latest version was this weekend around Friday/Saturday.

@log0ymxm I'll look into your YAML above and see if anything might have affected the codebase the last couple days.

**Comment by jackspirou** [comment]
Looks like there has been no updates since I last pulled the `drone:0.4` image:

![screen shot 2016-02-09 at 12 48 20 pm](https://cloud.githubusercontent.com/assets/5687406/12927008/856b1cca-cf2b-11e5-8728-3d0f1d3689f4.png)

**Comment by bradrydzewski** [comment]
@log0ymxm I bet that if you comment out the email step, it will work

I'm guessing that the email step is exiting with a failure, thus setting the local build state to failure, resulting in the subsequent hipchat notification to get marked as failed. 

If the email step is failing, you can run `docker logs drone` to see the drone server logs which should include failed notification logs.

**Comment by tboerger** [comment]
The email plugin can't work that way

**Comment by log0ymxm** [comment]
@bradrydzewski That makes sense. I just upgraded to 0.4, and did a quick port of the yml file. It looks like the smtp authentication has been moved into the plugin.

It does seem like unexpected behavior that a failed notify step would interfere with later notification steps.

**Comment by bradrydzewski** [comment]
@tboerger if this line of code is triggered in the email:
https://github.com/drone-plugins/drone-email/blob/7913a81e984ae98c7b117787548881ef8ad96c75/main.go#L56

then this line of code is triggered in the build runner:
https://github.com/drone/drone-exec/blob/master/runner/build.go#L127

this will set the build status to failed, which is reflected in subsequent plugins. I'm guessing this is why the hipchat plugin has a failure even though the build itself succeeds.

@log0ymxm agreed that this is unexpected behavior. Right now drone treats all plugins the same (deploy, publish, clone, etc) and will fail the build when the plugin fails. In all cases this is desirable except when dealing with notifications. We just need to put an `if` statement in place to ignore the error if the plugin is a notification plugin

**Comment by log0ymxm** [comment]
Removing the email notification fixes it. This bug should probably be closed or renamed if you plan on updating the notifications to not cascade errors.

**Comment by bradrydzewski** [comment]
@log0ymxm thanks for verifying the issue. Renaming this issue

**Comment by Jean85** [comment]
I had the same issue with the Slack plugin! Removing the email row from the YAML fixed id

**Comment by bradrydzewski** [comment]
I'm going to summarize this so I remember what happens next time I look at it.

When a notifications fails it returns a non-zero exit code which causes the build to get marked as failed. Subsequent notifications then have a failure status, even if the build itself passed. The permanent fix is that a non-zero exit code for notifications should be completely ignored.

This will be fixed in 0.6 which makes some slight changes to the yaml logic

**Comment by bradrydzewski** [comment]
the logic has been updated in 0.5 to be a bit more clear. Any failure to any build step will fail the build including notifications. With 0.5 drone no longer differentiates between a notification step and any other step -- a step is a step and a failure is a failure. The reason I'm comfortable making this change is because 0.5 now includes notification errors in the build logs, which was not the case in 0.4.
